### PR TITLE
GL: Set CanvasWidth/Height from input image size; tooltips, send output size 

### LINKED ISSUE #114: [Feature Request] Automatically set default width + height to the input image resolution
It would be nice to have a default width and height that matches the image you upload into the Controlnet media box. 

ie
 - I copy an image from discord, I don't know the resolution or I manually drag and drop or select an image to upload
 - I paste the image into the controlnet media upload box
 - Controlnet height and width parameters default to match the input image dimensions

### LINKED ISSUE #378: [Feature Request] Display uploaded image size to UI.  Workflow improvement
The more I use SD and ControlNet I'm finding that for every random picture I want to try, I first need to:

1. Open that picture in another photo editor
2. Write down or remember the picture dimensions
3. Use a calculator to compute the aspect ratio
4. Input the scaled version of that aspect ratio into the output image dimensions (in txt2img or img2img)

Steps 1 & 2 are in ControlNet.
Steps 3 & 4 are in base A1111.

Steps 1 & 2 can be eliminated if ControlNet simply reports the H & W of the uploaded image somewhere in the UI.

In playing with this there appears to be space just below the Image component to report this info in read-only Textboxes.

Possible implementation:

image

The Textboxes could be made even smaller, if Gradio supports that.

I'd be willing to take a crack at implementing this and filing a PR, but unfortunately, I'm from the desktop UI world and am apparently clueless on how the Gradio components and callbacks work together.  🙃

```
        input_image.orgpreprocess=input_image.preprocess
        input_image.preprocess=svgPreprocess
        input_image.upload(fn=get_image_dimensions, inputs=input_image, outputs=[])
```

I tried adding an "upload" callback to the input_image component but this ends up throwing errors from the svgPreprocess() method, which I'm guessing gets called first?

All that's needed is to get the image dimensions when an image is 'uploaded' or 'changed", perhaps.  Any guidance on making this work?  It seems rather simple, in theory.  😂

### PR DESCRIPTION
**Pull request by GeorgLegato**
Resolves #378 

# Video about latest button-rework 
(smaller, tooltips, near to the action)

https://user-images.githubusercontent.com/7210708/222928094-f3978652-233f-4bdb-896f-9afdf931fd73.mp4

# Proportional-Scale-Panel
image

# ToolTips! and "Send to Render size" implemented
image

# Quick-buttons to setup 512/768 proportional scaling, then click on "send to output size"
image

---
already merged these fixes from this branch:

Resolves #114 
image
image
image

### CODE REVIEW

File: scripts/controlnet.py [hunk 1]
```diff
@@ -342,16 +343,32 @@ def build_sliders(module):
...
+                        img = Image.open(buf)
+                        return [img.width,img.height]            
+
             def svgPreprocess(inputs):
                 if (inputs):
                     if (inputs['image'].startswith("data:image/svg+xml;base64,") and svgsupport):
                         svg_data = base64.b64decode(inputs['image'].replace('data:image/svg+xml;base64,',''))
                         drawing = svg2rlg(io.BytesIO(svg_data))
-                        png_data = renderPM.drawToString(drawing, fmt='PNG')
+                        png_data = renderPM.drawToString(drawing, fmt='PNG', dpi=shared.opts.data.get("control_net_cfg_svg_dpi"))
```
Comment (Mikubill) [inline scripts/controlnet.py:366]: always add default value for users don't want to restart WebUI

Comment (GeorgLegato) [reply scripts/controlnet.py]: assume the 2nd parameter in shared.opt.get is the default value. so it is fixed

**Comment by GeorgLegato** [comment]
@Mikubill lot of good changers here. size, tooltip... workflow improvements for the user. the zu design of the quickbuttons  (512/768) could be refined somehow

File: javascript/controlnet_hints.js [hunk 2]
```diff
@@ -0,0 +1,40 @@
+cnet_titles = {
+	"⏫": "Set this size as final processing output size",
+	"🔄": "Refresh",
+	"⇅": "Switch height/width",
+	"📷": "Mirror WebCam",
+	"⇄": "Leave WebCam"
```
Comment (missionfloyd) [inline javascript/controlnet_hints.js:6]:
```suggestion
	"📷": "Enable WebCam",
	"⇄": "Mirror WebCam"

```

File: scripts/controlnet.py [hunk 3]
```diff
@@ -343,29 +374,64 @@ def build_sliders(module):
...
+            with gr.Column(variant="panel"):
+                with gr.Box():
+                    with gr.Row():
+                        canvas_ratio =  gr.Slider(label="Canvas proportional scale %", minimum=1, maximum=1000, value=100, step=1, interactive=True)
+
+                    with gr.Row(variant="panel",):
+                        canvas_sw512 = ToolButton(value=set_w_512_symbol, label="W:512")
+                        canvas_sw768 = ToolButton(value=set_w_768_symbol, label="W:768")
+                        canvas_sh512 = ToolButton(value=set_h_512_symbol, label="H:512")
+                        canvas_sh768 = ToolButton(value=set_h_768_symbol, label="H:768")
```
Comment (missionfloyd) [inline scripts/controlnet.py:425]: > the design of the quickbuttons (512/768) could be refined somehow

They could be a bit wider.
```suggestion
                        canvas_sw512 = ToolButton(value=set_w_512_symbol, label="W:512", elem_id="btn_canvas_sw512")
                        canvas_sw768 = ToolButton(value=set_w_768_symbol, label="W:768", elem_id="btn_canvas_sw768")
                        canvas_sh512 = ToolButton(value=set_h_512_symbol, label="H:512", elem_id="btn_canvas_sh512")
                        canvas_sh768 = ToolButton(value=set_h_768_symbol, label="H:768", elem_id="btn_canvas_sh768")
```
Then add `sd-webui-controlnet/style.css`

```css
#btn_canvas_sw512, #btn_canvas_sw768, #btn_canvas_sh512, #btn_canvas_sh768 {
    max-width: 3.5em;
}
```

Comment (GeorgLegato) [reply scripts/controlnet.py]: thx, i add it

**Comment by GeorgLegato** [comment]
@Mikubill hi mate, since two weeks I am updating/merging this PR. any chance to get it merged? I see others working partially on similar topics, we could start new work instead of keeping our PR up todate, thx.

**Comment by Mikubill** [comment]
Too many buttons - to be honest. I suggest the following fix:
1. Move quick-buttons and scale-panel to somewhere like a floating window/iframe (at least, out of the main area).
2. Keep create canvas and preview button in their original position (for now).

**Comment by GeorgLegato** [comment]
> Too many buttons - to be honest. I suggest the following fix:
> 
> 1. Move quick-buttons and scale-panel to somewhere like a floating window/iframe (at least, out of the main area).
> 2. Keep create canvas and preview button in their original position (for now).

too much work, esp as now the ToOutput size is merged from LoricaMewinsky.
there are more buttons than before, yes, but less or same space used cause buttons contained long caption (due to no toolstips)
i close this PR, someone else shall continue/find concept by trial and error.

**Comment by Guilhermesscastro** [comment]
Hello, i stumbled upon your implementation and i really liked it @GeorgLegato, sorry for being a noob, but how can i use it ?

**Comment by GeorgLegato** [comment]
> Hello, i stumbled upon your implementation and i really liked it @GeorgLegato, sorry for being a noob, but how can i use it ?

Hi, my PR is not usable, we had concurrent ideas and implementations howto handle size/crop/scale/sendSize etc.
you could cherrypick some contetn from this PR, like the tooltip-stuff in the controlnet_hint.js. 
Meanwhile I think even SVG support is broken. who is using svg, right.
whatever you like here, feel free to create new PR based on those, for me it was too much effort and some kind of race for best implemention.

**Pull request closed without merging**